### DISCUSSION TITLE
Is Sqlite3 embedded in "DB Browser for SQLite"?

### DISCUSSION THREAD

**Discussion by wernerm** [sqlitebrowser/sqlitebrowser#3396] [General]
There seems to be a different version of Sqlite3 used by "DB Browser for SQLite" than on my system.

Does "DB Browser for SQLite" bundle and use its own version of Sqlite3, or does it use the underlying OS's installed version of Sqlite3?

Eg, running the application it on my Kubuntu desktop  works fine, but the `json1` extension does not seem to be available.

When I do a query on the command prompt:

```bash
$ sqlite3
sqlite> SELECT json_extract('{"foo":123,"animals":{"cat":"meowzer","dog":"woofles"}}', '$.animals.cat');
meowzer
sqlite> 
```

...I cannot do the same query in "DB Browser for SQLite"...

```sql
Execution finished with errors.
Result: no such function: json_extract
At line 1:
SELECT json_extract('{"foo":123,"animals":{"cat":"meowzer","dog":"woofles"}}', '$.animals.cat');
```

If this is the case (bundling Sqlite3 with the application), I would like to know where I can download the Sqlite extensions that are compiled into the Sqlite3 version of my OS.

**Comment by wernerm** [comment]
OK, I have found an answer for my first question by reading 
[Compiling DB4S on Ubuntu To Use A Pre Installed Version of SQLite](https://github.com/sqlitebrowser/sqlitebrowser/wiki/Compiling-DB4S-on-Ubuntu-To-Use-A-Pre-Installed-Version-of-SQLite)

That leaves me to wonder where I can get and install the `json1` extension for Ubuntu for DB4S?

**Comment by scottfurry** [comment]
A few factors to consider...
- how was DB4S installed? From context, I'm assuming this is from apt repository. Yes/No?
- If from apt, how was the app built. There is a build setting that allows for using the system sqlite3.
- Extensions can be loaded via the preferences.

**Reply by wernerm** [reply]
Yes, it's installed from the repo:

```bash
# deb-src https://ppa.launchpadcontent.net/linuxgndu/sqlitebrowser/ubuntu/ jammy main
```

Tbh I have installed it so long ago that I cannot even remember the build setting...

I was unfortunately not able to find the extension to download so it can be installed via preferences

**Reply by scottfurry** [reply]
Digging into the repository... It states sqlite3 as a dependency (currently 3.41 in Ubuntu) but does not defer to the system sqlite3. Which means it's using the internal sqlite3 package as part of our build. I can't remember, and I can't find, exact internal sqlite3 version at the moment (@chrisjlocke - help me out here, please).

You may have to uninstall and re-install from apt if its been awhile. FWIW, [apt cleaning](https://askubuntu.com/a/984800) is your friend. See [here](https://www.omgubuntu.co.uk/2016/08/5-ways-free-up-space-on-ubuntu).

**Reply by justinclift** [reply]
Yeah, it's a compile time option for SQLite these days. :smile:

**Reply by wernerm** [reply]
@scottfurry Thanks so much for looking into this. Things are starting to make sense.
The best way forward for us is probably just to switch to the Appimage, as @justinclift suggested.

**Comment by scottfurry** [comment]
As for `json1`, see [this](https://wiki.tcl-lang.org/page/SQLite+extension+JSON1). Seems the module use is so prolific that it's been pulled into the core library. See [this](https://askubuntu.com/questions/1200920/using-json1-extension-with-sqlite3) askubuntu page.

**Comment by justinclift** [comment]
Maybe try using the .AppImage build here?

https://github.com/sqlitebrowser/sqlitebrowser/releases/tag/continuous

That uses our latest development source code (at the moment it's pretty stable), and also includes the JSON extension. :smile:

**Reply by wernerm** [reply]
That's a good idea. Been using 3.12.99 on Linux for quite a while without any problems.

Tried the Appimage, I can confirm that `json` is included... but now I need to find out how to get SQLCipher working to open all of the encrypted databases... :facepalm:

**Comment by scottfurry** [comment]
@wernerm - I believe you want the app image nightly w/ SQLCipher
You can verify you have SQLCipher if it is listed on the About Dialog
![DB4S_About](https://github.com/sqlitebrowser/sqlitebrowser/assets/6686571/98453fee-07fe-4962-8ffd-df7ca551ee6d)

(@justinclift - I do not use appimages. Confirm SQLCipher is available in appimage?)

See these links for SQLCipher Usage:
- https://github.com/sqlcipher/sqlcipher (See README front page - examples are further down) 
- https://www.zetetic.net/sqlcipher/documentation/ (upstream user documentation)
- https://guardianproject.info/code/sqlcipher/ (another link w/ examples)

Good Luck

**Reply by justinclift** [reply]
Looks like the AppImage build only does SQLite at present:

https://github.com/sqlitebrowser/sqlitebrowser/blob/1ef78cb06aa5e490bffb99bc6767026d8a8a9e1d/.github/workflows/cppcmake.yml#L13-L67

On about line 32 of that it's grabbing the latest SQLite tarball for each build.  Doesn't seems to be any SQLCipher version of it.

Probably wouldn't be toooo hard to add one though.  Someone (not me in the short term) just needs to put in the time and effort to cut-n-paste-n-modify the SQLite section into a SQLCipher one.

---

@wernerm Not sure if it's of interest, but compiling our source code is pretty straight forward.

https://github.com/sqlitebrowser/sqlitebrowser/blob/master/BUILDING.md#build-with-sqlcipher-support
